scipy/sparse/coo.py

```diff
@@ -380,16 +380,22 @@ def todok(self):
         return dok
 
     def diagonal(self):
-        # Could be rewritten without the python loop.
-        # Data entries at the same (row, col) are summed.
-        n = min(self.shape)
-        ndata = self.data.shape[0]
-        d = np.zeros(n, dtype=self.dtype)
-        for i in xrange(ndata):
-            r = self.row[i]
-            if r == self.col[i]:
-                d[r] += self.data[i]
-        return d
+        diag_mask = self.row == self.col
+
+        # While it's tempting to use the following fast path when
+        # self.has_canonical_format, timing with random matrices shows that
+        # the speedup is marginal (even when the matrix is diagonal).
+        # diag[self.row[diag_mask]] = self.data[diag_mask]
```

Comment:
I'm not sure why we're avoiding the tempting fast path here. It seems like this wouldn't add much complexity, and it would definitely save on memory if not time:
```python
if self.has_canonical_format:
    row, data = self.row[diag_mask], self.data[diag_mask]
else:
    row, _, data = self._sum_duplicates(...)
```